### ISSUE TITLE
[BUG] Compute deadlock: loop-carried recurrence + compiler-managed intermediate DFBs

### ISSUE THREAD

**Issue by brnorris03** [tenstorrent/tt-lang#666]
## Summary

A `ttl.compute` kernel that combines a **loop-carried recurrence** (e.g. a
running `max` updated each iteration) with **compiler-managed intermediate
values** (matmul / reduce / transpose results left as SSA, materialized by
`ttl-insert-intermediate-dfbs`) **deadlocks on device**. Removing only the
loop-carried recurrence — keeping the same matmul/reduce/exp/transpose SSA
chain — makes the identical kernel pass.

Reproduces on `main` (commit 8eb21608) on Blackhole.

## Reproducers (complete source)

Single file, two ops selectable via the `REPRO` env var. `v2_reuse` (no
loop-carried state) **PASSES**; `v3_state` (= v2 **plus** one running-max
recurrence `s = max(s_old, reduce_max(qk))`) **HANGS**. That single added
recurrence is the entire delta.

```python
# _repro_grow.py
import os
import torch
import ttl
import ttnn

TILE = 32
N_ITERS = 4
K = 2
M = 2


def to_dram(dev, t):
    return ttnn.from_torch(
        t.contiguous(), dtype=ttnn.bfloat16, layout=ttnn.TILE_LAYOUT,
        device=dev, memory_config=ttnn.DRAM_MEMORY_CONFIG,
    )


@ttl.operation(grid=(1, 1))
def v2_reuse(a, b, c, out):
    a_cb = ttl.make_dataflow_buffer_like(a, shape=(1, K), block_count=2)
    b_cb = ttl.make_dataflow_buffer_like(b, shape=(M, K), block_count=2)
    qk_cb = ttl.make_dataflow_buffer_like(a, shape=(1, M), block_count=2)
    mx_cb = ttl.make_dataflow_buffer_like(a, shape=(1, 1), block_count=2)
    es_cb = ttl.make_dataflow_buffer_like(a, shape=(1, 1), block_count=2)
    out_cb = ttl.make_dataflow_buffer_like(out, shape=(1, 1), block_count=N_ITERS)

    @ttl.compute()
    def compute():
        for _ in range(N_ITERS):
            a_blk = a_cb.wait()
            b_blk = b_cb.wait()
            qk_w = qk_cb.reserve(); qk_w.store(a_blk @ ttl.transpose(b_blk))
            qk = qk_cb.wait()
            mx_w = mx_cb.reserve(); mx_w.store(ttl.math.reduce_max(qk, dims=[1]))
            mx = mx_cb.wait()
            mx_bc = ttl.block.broadcast(mx, dims=[1], shape=(1, M))
            ex = ttl.exp(ttl.sub(qk, mx_bc))
            es_w = es_cb.reserve(); es_w.store(ttl.math.reduce_sum(ex, dims=[1]))
            es = es_cb.wait()
            ow = out_cb.reserve(); ow.store(es)

    @ttl.datamovement()
    def dm():
        for _ in range(N_ITERS):
            ad = a_cb.reserve(); ttl.copy(a[0:1, 0:K], ad)
            bd = b_cb.reserve(); ttl.copy(b[0:M, 0:K], bd)
        for _ in range(N_ITERS):
            ob = out_cb.wait(); ttl.copy(ob, out[0:1, 0:1])

    @ttl.datamovement()
    def dm_unused():
        pass


@ttl.operation(grid=(1, 1))
def v3_state(a, b, c, out):
    a_cb = ttl.make_dataflow_buffer_like(a, shape=(1, K), block_count=2)
    b_cb = ttl.make_dataflow_buffer_like(b, shape=(M, K), block_count=2)
    qk_cb = ttl.make_dataflow_buffer_like(a, shape=(1, M), block_count=2)
    mx_cb = ttl.make_dataflow_buffer_like(a, shape=(1, 1), block_count=2)
    es_cb = ttl.make_dataflow_buffer_like(a, shape=(1, 1), block_count=2)
    st_cb = ttl.make_dataflow_buffer_like(a, shape=(1, 1), block_count=2)  # loop-carried state
    out_cb = ttl.make_dataflow_buffer_like(out, shape=(1, 1), block_count=N_ITERS)

    @ttl.compute()
    def compute():
        s0 = st_cb.reserve(); s0.store(ttl.block.fill(-1e30, shape=(1, 1)))
        for _ in range(N_ITERS):
            a_blk = a_cb.wait()
            b_blk = b_cb.wait()
            qk_w = qk_cb.reserve(); qk_w.store(a_blk @ ttl.transpose(b_blk))
            qk = qk_cb.wait()
            mx_w = mx_cb.reserve(); mx_w.store(ttl.math.reduce_max(qk, dims=[1]))
            mx = mx_cb.wait()
            s_old = st_cb.wait()                          # <-- recurrence
            s_new = ttl.math.max(s_old, mx)               # <-- recurrence
            s_next = st_cb.reserve(); s_next.store(s_new)  # <-- recurrence
            mx_bc = ttl.block.broadcast(s_new, dims=[1], shape=(1, M))
            ex = ttl.exp(ttl.sub(qk, mx_bc))
            es_w = es_cb.reserve(); es_w.store(ttl.math.reduce_sum(ex, dims=[1]))
            es = es_cb.wait()
            ow = out_cb.reserve(); ow.store(es)
        _ = st_cb.wait()

    @ttl.datamovement()
    def dm():
        for _ in range(N_ITERS):
            ad = a_cb.reserve(); ttl.copy(a[0:1, 0:K], ad)
            bd = b_cb.reserve(); ttl.copy(b[0:M, 0:K], bd)
        for _ in range(N_ITERS):
            ob = out_cb.wait(); ttl.copy(ob, out[0:1, 0:1])

    @ttl.datamovement()
    def dm_unused():
        pass


def main():
    which = os.getenv("REPRO", "v2_reuse")
    fn = {"v2_reuse": v2_reuse, "v3_state": v3_state}[which]
    dev = ttnn.open_device(device_id=0)
    try:
        a = to_dram(dev, torch.randn(TILE, K * TILE, dtype=torch.bfloat16))
        b = to_dram(dev, torch.randn(M * TILE, K * TILE, dtype=torch.bfloat16))
        c = to_dram(dev, torch.randn(M * TILE, TILE, dtype=torch.bfloat16))
        out = to_dram(dev, torch.zeros(TILE, TILE, dtype=torch.bfloat16))
        print(f"running {which}...")
        fn(a, b, c, out)
        ttnn.synchronize_device(dev)
        print(f"REPRO {which} PASS")
    finally:
        ttnn.close_device(dev)


if __name__ == "__main__":
    main()
```

Run:

```
REPRO=v2_reuse python _repro_grow.py   # prints "REPRO v2_reuse PASS"
REPRO=v3_state python _repro_grow.py   # hangs (no output after "Compiled kernel ready")
```

## Observed behavior

- `v2_reuse`: completes, `REPRO v2_reuse PASS`.
- `v3_state`: kernels compile and dispatch, then the device hangs forever
  (with `TT_METAL_WATCHER=10` the device keeps being polled; no progress).

## Evidence / analysis

- **Not a circular-buffer logic bug.** The IR after `ttl-insert-cb-sync` is
  balanced: every `cb_reserve`/`cb_push` and `cb_wait`/`cb_pop` pairs up
  correctly across the producer/consumer threads.
- **Not back-pressure depth.** Raising the cross-thread DFB `block_count`
  (k/v) from 2 to 4 does **not** fix it.
- **Watcher** (Blackhole) on the single active worker core shows the
  data-movement RISC (NCRISC) blocked at `CRBW` (`cb_reserve_back` wait); the
  compute TRISC waypoints are at *Done* markers (`UABD`, `MWDD`), i.e. the
  compute pipeline is stalled and back-pressures the producer.
- After `ttl-assign-dst`, the loop body is split into many `ttl.dst_section`s
  (one per compute op). The recurrence adds a value (`s_new`) that is
  multi-used — materialized into a compiler-allocated DFB for the broadcast
  **and** written back to the state DFB — which appears to tip the
  DST-section / unpack-math-pack scheduling into a deadlocking order.

Conclusion: the deadlock is in DST-register / `ttl.dst_section` scheduling
(`ttl-assign-dst` -> `ttl-lower-to-loops`), surfaced by combining a
loop-carried recurrence with compiler-managed intermediate DFBs.

## Workaround

Manually stage **every** intermediate in its own single-use DFB instead of
relying on `ttl-insert-intermediate-dfbs` (this keeps DST live-ranges short).
A fully hand-staged flash-attention `flash_shard_kd` (~25 DFBs) runs correctly;
reducing it to rely on compiler-managed intermediates reintroduces this hang.

## Impact

Blocks authoring minimal-DFB online-softmax / flash-attention-style kernels:
any loop-carried recurrence (running max/sum/output) combined with
compiler-materialized op intermediates can deadlock, forcing users to
hand-stage all intermediates.

**Comment by brnorris03** [comment]
## Root cause located: `convert-ttl-to-compute` inverts push/wait for a multi-output intermediate DFB

Traced the deadlock to the `convert-ttl-to-compute` pass
(`lib/Dialect/TTL/Transforms/ConvertTTLToCompute.cpp`). The file is identical
on `main`, so the analysis applies to `main`.

### Mechanism

The running-max recurrence makes the `max` result feed two consumers in the
same step: the broadcast (through a compiler-allocated intermediate DFB) and
the loop-carried state DFB. `convert-ttl-to-compute` fuses this into a single
multi-output `ttl.compute` that writes both DFBs.

While forming that compute, the pass relocates the intermediate DFB's
`cb_push` to after the compute (so the push runs after `pack_tile` writes the
data). That relocation is correct on its own, but the consumer-side `cb_wait`
that the frontend emits immediately after the push (the `reserve -> store ->
push -> wait` materialization pattern) is left in place, before the compute.
The result is that `cb_wait` precedes `cb_push` for that intra-thread DFB.

### Evidence

Per-pass ordering of the intermediate DFB's circular-buffer ops (cb_index 9 in
the reproducer):

```
... -> convert-ttl-to-compute (input):  cb_reserve, cb_push, cb_wait, cb_pop   (correct)
       convert-ttl-to-compute (output): cb_reserve, cb_wait, cb_push, cb_pop   (inverted)
```

The inversion first appears at the output of `convert-ttl-to-compute` and
persists to the emitted compute kernel, where `cb_ctargN.wait_front(...)`
appears before the matching `cb_ctargN.push_back(...)`.

`v2_reuse` (no recurrence) keeps `cb_push` before `cb_wait` and passes;
`v3_state` (adds the recurrence) produces the inverted order and hangs.

### Why it deadlocks on device

For an intra-thread DFB the consumer wait must follow the producer push. With
the order inverted, the three compute RISCs form a wait-for cycle:

- UNPACK blocks at the intermediate DFB's `cb_wait_front` (the buffer has not
  been pushed yet), so it never reaches the `copy_tile` that unpacks the
  operands of the producing compute.
- PACK cannot `cb_push_back` the intermediate DFB until it packs the compute
  result, which requires MATH to commit.
- MATH cannot commit until UNPACK provides the operands.

The data-movement thread then blocks at `cb_reserve_back` (watcher waypoint
`CRBW`) behind the stalled compute, which is the externally visible symptom.

### Fix (implemented; verification in progress)

In `ConvertTTLToCompute.cpp`, when relocating a `cb_push` past the compute,
also move the stranded consumer `cb_wait` (and its `attach_cb`) on the same
DFB to follow the relocated push, restoring push-before-wait. Verifying on
device against the reproducer and the originating flash-attention kernel.
